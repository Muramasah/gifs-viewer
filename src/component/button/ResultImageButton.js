import React from 'react';
import IsFavoriteButton from './IsFavoriteButton';
import LikeButton from './LikeButton';

const ResultImageButton = ({ gif, onClickButton }) => {
    if (gif.isFavorite) {
        return <IsFavoriteButton />
    } else {
        return <LikeButton gif={gif} onClickButton={() => { onClickButton(gif) }} />
    }
}

export default ResultImageButton;