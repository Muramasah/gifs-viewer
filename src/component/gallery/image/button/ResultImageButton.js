import React from 'react';
import IsFavoriteButton from './IsFavoriteButton';
import LikeButton from './LikeButton';

const ResultImageButton = ({ isFavorite, onLike }) => {
    if (isFavorite) {
        return <IsFavoriteButton />
    } else {
        return <LikeButton onClick={onLike} />
    }
}

export default ResultImageButton;