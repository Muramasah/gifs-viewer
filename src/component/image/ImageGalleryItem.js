//Library
import React, { Component } from 'react';
//Component
import LoadingImage from './LoadingImage';
import MiddleInlineDiv from './presenter/MiddleInlineDiv';
import StyledImageItem from './presenter/StyledImageItem';
//Helper
import { showElementByID, hideElementByID } from '../../helper/dom';

/**
 * Component witch represents the item showed by a ItemListView component.
 * uses the state to handle the changes in the src of the gifs and the
 * loading images.
 */
class ImageGalleryItem extends Component {
    /**
     * Constructor method
     * @param {*} props Properties of the component
     * @param {string} props.src Gif url
     * @param {string} props.uiID Interface ID generted bi ItemListView component
     * @param {string} props.alt Alt text for the image, extracted from the gif slug.
     * @param {component} props.ItemButton Component passed by the specific type ImageItem component,
     * such as TrendinImageItem, ResultImageItem and FavoriteImageItem.
     * @param {object} props.gif referense to de gif.
     * @param {function} props.onClickButton Function to execute when the user clicks on the ImageButton
     * component.
     * @returns {undefined} void
     */
    constructor(props) {
        super(props)

        this.state = {
            src: props.src,
            uiID: props.uiID
        }
    }
    /**
     * @see https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops
     * @param {object} newProps New properties passed to de component 
     * @param {*} prevState
     * @returns {object} if there are properties that has changed, returns the new state
     * @returns {null} if the properties has not changed, returns null
     */
    static getDerivedStateFromProps({ uiID, src }, prevState) {
        if (src !== prevState.src) {
            hideElementByID(prevState.uiID);
            showElementByID('loading_' + prevState.uiID);

            return { src, uiID }
        } else {
            return null;
        }
    }

    /**
     * When the image is loaded shows the image element and hides the loading one.
     * Also executes a scroll to de bottom of the page.
     * @param {object} event accepts react event object, but is not used.
     * @returns {undefined} void 
     */
    onLoad() {
        const { uiID } = this.state;

        showElementByID(uiID);
        hideElementByID('loading_' + uiID);
        window.scrollTo(0, document.body.scrollHeight);
    }

    /**
     * When an image is clicked opens a new window with the image in full size.
     * @param {object} event accepts react event object, is used to obtain the src
     * of the image.
     * @returns {undefined} void 
     */
    onClick(event) {
        window.open(
            event.target.src,
            'newwindow',
            'width=600,height=400'
        );
    }

    render() {
        const { alt, ItemButton, gif, onClickButton } = this.props;
        const { uiID, src } = this.state;

        return (
            <MiddleInlineDiv>
                <div id={uiID} style={{ display: 'none' }}>
                    <StyledImageItem
                        src={src}
                        alt={alt}
                        onLoad={this.onLoad.bind(this)}
                        onClick={this.onClick.bind(this)}
                    />
                    {
                        ItemButton
                            ? <ItemButton
                                gif={gif}
                                onClickButton={onClickButton}
                            />
                            : false
                    }
                </div>

                <LoadingImage id={'loading_' + uiID} />
            </MiddleInlineDiv>
        );
    }
}

export default ImageGalleryItem;