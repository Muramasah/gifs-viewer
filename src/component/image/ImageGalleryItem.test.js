import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ImageGalleryItem from './ImageGalleryItem';
import ResultImageButton from '../button/ResultImageButton';

configure({ adapter: new Adapter() });

jest.mock('../../helper/dom'); // this happens automatically with automocking

const showElementByID = require('../../helper/dom').showElementByID;
showElementByID.mockImplementation(() => '');

const hideElementByID = require('../../helper/dom').hideElementByID;
hideElementByID.mockImplementation(() => '');

const gif = {
    id: '3ohc123V1Dr1nf8lcQ',
    slug: 'dance-cute-bear-3ohc123V1Dr1nf8lcQ',
    url: 'https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif'
}

test('ImageGalleryItem renders with ResultImageButton without crashing', () => {
    const imageGalleryItem = shallow(
        <ImageGalleryItem
            src={'https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif'}
            uiID={'poU45wTKf1Qut1XirnOq'}
            alt={'alt'}
            ItemButton={ResultImageButton}
            gif={gif}
            onClickButton={() => { }}
        />
    );

    expect(imageGalleryItem).toMatchSnapshot();
});

test('getDerivedStateFromProps', () => {
    const prevState={
        uiID:'3ohc123V1Dr1nf8lcQ',
        src:'https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif'
    }
    const newProps = {
        uiID:'poU45wTKf1Qut1XirnOq',
        src:'https://media0.giphy.com/media/QQHBQZnIGEU8g/giphy.gif'
    };

    const derivatedState = ImageGalleryItem.getDerivedStateFromProps(newProps, prevState)

    expect(derivatedState.uiID).toBe(newProps.uiID);
    expect(derivatedState.src).toBe(newProps.src);
});

test('onLoad', ()=>{
    const imageGalleryItem = shallow(
        <ImageGalleryItem
            src={'https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif'}
            uiID={'poU45wTKf1Qut1XirnOq'}
            alt={'alt'}
            ItemButton={ResultImageButton}
            gif={gif}
            onClickButton={() => { }}
        />
    );

    imageGalleryItem.instance().onLoad();
    expect(imageGalleryItem).toMatchSnapshot();
})