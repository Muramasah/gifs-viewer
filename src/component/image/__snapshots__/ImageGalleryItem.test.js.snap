// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGalleryItem renders with ResultImageButton without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImageGalleryItem
    ItemButton={[Function]}
    alt="alt"
    gif={
        Object {
            "id": "3ohc123V1Dr1nf8lcQ",
            "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
            "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
          }
    }
    onClickButton={[Function]}
    src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
    uiID="poU45wTKf1Qut1XirnOq"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div
          id="poU45wTKf1Qut1XirnOq"
          style={
                    Object {
                              "display": "none",
                            }
          }
>
          <styled.img
                    alt="alt"
                    onClick={[Function]}
                    onLoad={[Function]}
                    src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
          />
          <ResultImageButton
                    gif={
                              Object {
                                        "id": "3ohc123V1Dr1nf8lcQ",
                                        "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                        "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                      }
                    }
                    onClickButton={[Function]}
          />
</div>,
        <LoadingImage
          id="loading_poU45wTKf1Qut1XirnOq"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <styled.img
              alt="alt"
              onClick={[Function]}
              onLoad={[Function]}
              src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
/>,
            <ResultImageButton
              gif={
                            Object {
                                          "id": "3ohc123V1Dr1nf8lcQ",
                                          "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                          "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                        }
              }
              onClickButton={[Function]}
/>,
          ],
          "id": "poU45wTKf1Qut1XirnOq",
          "style": Object {
            "display": "none",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "alt": "alt",
              "onClick": [Function],
              "onLoad": [Function],
              "src": "https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "gif": Object {
                "id": "3ohc123V1Dr1nf8lcQ",
                "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
              },
              "onClickButton": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "id": "loading_poU45wTKf1Qut1XirnOq",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div
            id="poU45wTKf1Qut1XirnOq"
            style={
                        Object {
                                    "display": "none",
                                  }
            }
>
            <styled.img
                        alt="alt"
                        onClick={[Function]}
                        onLoad={[Function]}
                        src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
            />
            <ResultImageButton
                        gif={
                                    Object {
                                                "id": "3ohc123V1Dr1nf8lcQ",
                                                "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                                "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                              }
                        }
                        onClickButton={[Function]}
            />
</div>,
          <LoadingImage
            id="loading_poU45wTKf1Qut1XirnOq"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <styled.img
                alt="alt"
                onClick={[Function]}
                onLoad={[Function]}
                src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
/>,
              <ResultImageButton
                gif={
                                Object {
                                                "id": "3ohc123V1Dr1nf8lcQ",
                                                "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                                "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                              }
                }
                onClickButton={[Function]}
/>,
            ],
            "id": "poU45wTKf1Qut1XirnOq",
            "style": Object {
              "display": "none",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "alt": "alt",
                "onClick": [Function],
                "onLoad": [Function],
                "src": "https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "gif": Object {
                  "id": "3ohc123V1Dr1nf8lcQ",
                  "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                  "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                },
                "onClickButton": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "id": "loading_poU45wTKf1Qut1XirnOq",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`onLoad 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImageGalleryItem
    ItemButton={[Function]}
    alt="alt"
    gif={
        Object {
            "id": "3ohc123V1Dr1nf8lcQ",
            "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
            "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
          }
    }
    onClickButton={[Function]}
    src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
    uiID="poU45wTKf1Qut1XirnOq"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div
          id="poU45wTKf1Qut1XirnOq"
          style={
                    Object {
                              "display": "none",
                            }
          }
>
          <styled.img
                    alt="alt"
                    onClick={[Function]}
                    onLoad={[Function]}
                    src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
          />
          <ResultImageButton
                    gif={
                              Object {
                                        "id": "3ohc123V1Dr1nf8lcQ",
                                        "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                        "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                      }
                    }
                    onClickButton={[Function]}
          />
</div>,
        <LoadingImage
          id="loading_poU45wTKf1Qut1XirnOq"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <styled.img
              alt="alt"
              onClick={[Function]}
              onLoad={[Function]}
              src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
/>,
            <ResultImageButton
              gif={
                            Object {
                                          "id": "3ohc123V1Dr1nf8lcQ",
                                          "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                          "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                        }
              }
              onClickButton={[Function]}
/>,
          ],
          "id": "poU45wTKf1Qut1XirnOq",
          "style": Object {
            "display": "none",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "alt": "alt",
              "onClick": [Function],
              "onLoad": [Function],
              "src": "https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "gif": Object {
                "id": "3ohc123V1Dr1nf8lcQ",
                "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
              },
              "onClickButton": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "id": "loading_poU45wTKf1Qut1XirnOq",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div
            id="poU45wTKf1Qut1XirnOq"
            style={
                        Object {
                                    "display": "none",
                                  }
            }
>
            <styled.img
                        alt="alt"
                        onClick={[Function]}
                        onLoad={[Function]}
                        src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
            />
            <ResultImageButton
                        gif={
                                    Object {
                                                "id": "3ohc123V1Dr1nf8lcQ",
                                                "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                                "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                              }
                        }
                        onClickButton={[Function]}
            />
</div>,
          <LoadingImage
            id="loading_poU45wTKf1Qut1XirnOq"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <styled.img
                alt="alt"
                onClick={[Function]}
                onLoad={[Function]}
                src="https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif"
/>,
              <ResultImageButton
                gif={
                                Object {
                                                "id": "3ohc123V1Dr1nf8lcQ",
                                                "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                                                "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                                              }
                }
                onClickButton={[Function]}
/>,
            ],
            "id": "poU45wTKf1Qut1XirnOq",
            "style": Object {
              "display": "none",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "alt": "alt",
                "onClick": [Function],
                "onLoad": [Function],
                "src": "https://media0.giphy.com/media/11Ldh8hMMCEkta/giphy.gif",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "gif": Object {
                  "id": "3ohc123V1Dr1nf8lcQ",
                  "slug": "dance-cute-bear-3ohc123V1Dr1nf8lcQ",
                  "url": "https://media1.giphy.com/media/3ohc123V1Dr1nf8lcQ/giphy.gif",
                },
                "onClickButton": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "id": "loading_poU45wTKf1Qut1XirnOq",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
